libvast/src/table_slice.cpp

```diff
@@ -823,4 +815,266 @@ auto resolve_operand(const table_slice& slice, const operand& op)
+    = transform_columns(field.type, struct_array, transformations);
+  VAST_ASSERT(output_type);
+  VAST_ASSERT(output_struct_array);
+  auto result = result_type{};
+  result.reserve(output_struct_array->num_fields());
+  const auto& output_rt = caf::get<record_type>(output_type);
+  for (int i = 0; i < output_struct_array->num_fields(); ++i) {
+    const auto field_view = output_rt.field(i);
+    result.push_back({
+      {std::string{field_view.name}, field_view.type},
+      output_struct_array->field(i),
+    });
+  }
+  return result;
+}
+
+auto make_flatten_transformation(
+  std::string_view separator, const std::string& name_prefix,
+  std::vector<std::shared_ptr<arrow::Array>> list_offsets)
+  -> indexed_transformation::function_type {
+  return [=](struct record_type::field field,
+             std::shared_ptr<arrow::Array> array)
+           -> std::vector<std::pair<struct record_type::field,
+                                    std::shared_ptr<arrow::Array>>> {
+    using result_type = std::vector<
+      std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>>;
+    auto f = detail::overload{
+      [&]<concrete_type Type>(const Type&) -> result_type {
+        // Return unchanged, but use prefix, and wrap in a list if we need to.
+        if (list_offsets.empty()) {
+          return {
+            {
+              {
+                fmt::format("{}{}", name_prefix, field.name),
+                field.type,
+              },
+              array,
+            },
+          };
+        }
+        auto combined_offsets = combine_offsets(list_offsets);
+        return {
+          {
+            {
+              fmt::format("{}{}", name_prefix, field.name),
+              type{list_type{field.type}},
+            },
+            arrow::ListArray::FromArrays(*combined_offsets, *array).ValueOrDie(),
+          },
+        };
+      },
+      [&](const list_type&) -> result_type {
+        return flatten_list(separator, name_prefix, list_offsets,
+                            std::move(field), array);
+      },
+      [&](const record_type&) -> result_type {
+        return flatten_record(separator, name_prefix, list_offsets,
+                              std::move(field), array);
+      },
+    };
+    auto result = caf::visit(f, field.type);
+    return result;
+  };
+}
+
+auto make_rename_transformation(std::string new_name)
+  -> indexed_transformation::function_type {
+  return [new_name = std::move(new_name)](struct record_type::field field,
+                                          std::shared_ptr<arrow::Array> array)
+           -> std::vector<std::pair<struct record_type::field,
+                                    std::shared_ptr<arrow::Array>>> {
+    return {
+      {
+        {
+          new_name,
+          field.type,
+        },
+        array,
+      },
+    };
+  };
+}
+
+} // namespace
+
+auto flatten(table_slice slice, std::string_view separator) -> flatten_result {
```

Comment:
The question boils down to whether we want to treat top-level nulls for list values separately from nested nulls for list values. I am honestly not sure what the right approach for this is.

Reply:
Well, in this case we have an `unset` value vs. a set value that is `empty`. How differently does Zeek handle those types internally? How strong is the difference between these two value types for Zeek users? 

Or even for VAST users: Adding  something like `where answers` to the pipeline above after `flatten`ing would yield additional events that, in terms of the original information they contained, shouldn't be yielded (events where `null` has been transformed to non-null empty values).

If we can be idiomatic, we could say that we could remove non-existent values (e.g. nulls in lists), but not replace non-existent values with existent, but not useful values (e.g. `null` -> empty list).

Reply:
Why would we touch a list value that's `null`? Flattening should only change the structure, not the values. 

> Well, in this case we have an `unset` value vs. a set value that is `empty`. How differently does Zeek handle those types internally? How strong is the difference between these two value types for Zeek users?

These are different, in the same way that the empty string is different from null.

Reply:
I thought about this over the weekend a lot. Here's the list flattening behavior that I think we may want:

1. We flatten into `null` if the outermost list is `null`.
2. We keep the null values of the innermost list.
3. We remove null values of all lists except for the innermost list.

This would give the following behavior for a `list<list<uint64>>`:

|original<br>`list<list<uint64>>`|flattened<br>`list<uint64>`|
|:-:|:-:|
|`null`|`null`|
|`[null]`|`[]`|
|`[[null]]`|`[null]`|
|`[[null, 1], null, [2, null, 3]`|`[null, 1, 2, null, 3]`|

I'd like to agree on this behavior (or a different one, for that matter) before we make any further changes to this PR. Wdyt @Dakostu @mavam?

Reply:
Compare this to JavaScript:

```javascript
[null].flat()
// Output: [ null ]
[[null]].flat()
// Output: [ null ]
[[null, 1], null, [2, null, 3]].flat()
// Output: [ null, 1, null, 2, null, 3 ]
```

This makes more sense to me than your proposal that for some to me unclear reason strips null values (which are valid list elements).

Reply:
The problem with that is that you're treating nulls differently depending on the nesting:

1. In `[[null: uint64]]`, the `null` is treated like `null: uint64`
2. In `[null: list<uint64>]`, the `null` is treated like `[null]: list<uint64>`.
3. In `null: list<list<uint64>>`, the `null` is treated like `null: list<list<uint64>>`.

I find the difference between (2) and (3) very unintuitive, as in one case the `null` value denotes the absence of a list, and in the other it silently converts into a list with a single `null` element. We should look closer at what other Arrow-based systems do, rather than languages with known-to-be-inconsistent type systems like JS.

Reply:
To me, the main point still holds: `null` is a regular list element (regardless of its type) and should not be removed by a structural transformation like `flatten`.

Reply:
I skimmed through some of the [projects powered by Arrow](https://arrow.apache.org/powered_by/) and the only flattening I could find that is not Python- , JS- or Rust-based is [FlattenArrays() in Cylon](https://github.com/cylondata/cylon/blob/d9a64204f9237729694df869b448fb67eed1440c/cpp/src/cylon/util/flatten_array.cpp#L289). But that function works slightly differently to our flattening; it basically [flattens multiple same-length arrays into one](https://github.com/cylondata/cylon/blob/30bc9f3b51931adca95886342e184678e6b42d77/cpp/src/cylon/util/flatten_array.hpp#L50), [combining elements](https://github.com/cylondata/cylon/blob/30bc9f3b51931adca95886342e184678e6b42d77/cpp/test/flatten_array_test.cpp#L103) and [adding empty elements in lieu of any `null` value](https://github.com/cylondata/cylon/blob/30bc9f3b51931adca95886342e184678e6b42d77/cpp/src/cylon/util/flatten_array.hpp#L57).

While I get that removing elements from flattening may seem unintuitive: The flattening of a multi-dimensional `list<list<T>` into `list<T>`  - to me - signifies that there shouldn't be any `list<T>` values in the list anymore. This also includes `null` values that denoted the absence of lists. This value that denotes absence would be "flattened" into non-existence, since the `list<T>`, in terms of dimensionality, should not be intelligent enough anymore to know about `list<list<T>>` values.

To cut a long story short: 
* `[[null]]` in `list<list<T>>` = "No  `T` value" -> flattened into `null` in `list<T>`
* `[null]` in `list<list<T>>` = "No list" -> flattened into non-existence in `list<T>`

That mental model makes sense to me, at least in a strongly typed world.

Reply:
I do have a strong opinion in the other direction: we should avoid making a design decision that does not interact well with the libraries we are using, so I think we should use the exact behavior of Arrow's list flattening function here to avoid an inconsistent behavior when we call into some Arrow functionality in the future that flattens for us implicitly.

It appears that we are stuck here then in this discussion, so I suggest we take this into a call because this thread isn't an effective use of our time.

Reply:
If you think that both options are valuable, then make this a configuration option. I have use cases where I *don't* my list cardinality to change during analysis (counting elements is one of them). If you have others where this is desired, it's not that hard to make everyone happy.

If that's not feasible, we need a call :-).

Reply:
We discussed this in person and agreed on the following: By default, flatten both lists and records with the Arrow-default semantics. Add an option that causes the operator to only flatten records with the Arrow-default semantics, but not lists.

I will update this PR sometime next week to implement this.

Reply:
For the sake of providing basic flattening functionality, we will ship `flatten` as-is and wait for Alpha test feedback.